bosk-mongo/DEVELOPERS.md

```diff
@@ -0,0 +1,246 @@
+which rests on a kind of framework shell object called `MainDriver` that deals with
+initialization, error handling, logging, change stream management, database transactions, and coordination between threads.
+When anything goes awry, `MainDriver` can discard and replace the `FormatDriver` object,
+which allows `FormatDriver` to be largely ignorant of fault tolerance concerns:
+`FormatDriver` can simply throw exceptions when errors occur, and let `MainDriver` to handle the recovery operations.
+
+The `MainDriver` functionality is further divided into three areas:
+- the change stream cursor lifecycle is managed by `ChangeReceiver`, which houses the background thread that executes all change stream operations and event handling;
+- the `BoskDriver` `submit` methods are implemented by `MainDriver` itself; and
+- the `MainDriver.Listener` connects `ChangeReceiver` to `MainDriver` so `MainDriver` can respond to any errors that occur on the background thread.
+
+Whenever possible, all error conditions handled internally are signaled by checked exceptions,
+meaning the Java compiler can help check that we have handled them all.
+In general, we do not shy away from defining new types of exceptions,
+erring on the side of adding a new exception for clarity rather than reusing an existing one.
+
+Testing is performed by JUnit 5 tests using Testcontainers and Toxiproxy.
+Testcontainers allows us to verify our database interactions against an actual MongoDB database,
+while Toxiproxy allows us to induce various network errors and verify our response to them.
+(Mocks are not generally suitable for testing error cases,
+since the precise exceptions thrown by the client library are not well documented,
+and since they are runtime exceptions, we get no help from the Java compiler.)
+We have developed a few of our own JUnit 5 extensions to make this testing more convenient,
+such as `@ParametersByName` and `@DisruptsMongoService`.
+
+### TODO: Points to cover
+(This document is a work in progress.)
+
+- Basic operation
+- Reads always come from memory. That's always true in Bosk, and a driver can't change that even if it wanted to.
+- Driver operations lead to database operations; they are not forwarded downstream
+	- Translated from bosk to MongoDB by a `FormatDriver`
+	- `initialRoot` is a special case that will need to be documented separately
+- Change events bring the database operations back to `MongoDriver` via `ChangeReceiver`, which forwards them to `FormatDriver`
+	- From there, they are translated from MongoDB back to Bosk and forwarded to the downstream driver
+- With a single bosk interacting with the database, all of this makes little difference, but
+	- You get persistence/durability
+	- You get replication for free!
+- The cursor lifecycle
+- Two different lifetimes: permanent and cursor
+- Responsibilities of the background thread
+- `FormatDriver`
+- Division of responsibilities from the draft Principles of Operation doc
```

Comment:
Is this publicly available anywhere? I didn't see it in the docs directory.

Reply:
I'll share it with you.